
syntax = "proto3";

package mimir;

service Coordinator {
  rpc RegisterParty(PartyInfo) returns (RegisterAck);
  rpc CreateSession(SessionSpec) returns (SessionAck);
  rpc SubmitPrompt(SubmitPromptReq) returns (SubmitPromptAck);
  rpc RunInference(RunInferenceReq) returns (stream InferenceEvent);
  rpc FetchMetrics(Empty) returns (Metrics);
}

message Empty {}

message PartyInfo {
  string party_id = 1;
  string addr = 2;
}

message RegisterAck {
  bool ok = 1;
  string msg = 2;
}

message SessionSpec {
  string session_id = 1;
  string model_name = 2;
  string mpc_ring = 3; // e.g., "2^64"
}

message SessionAck {
  bool ok = 1;
  string msg = 2;
}

message SubmitPromptReq {
  string session_id = 1;
  string owner_party = 2;
  bytes token_ids = 3; // serialized int64 array for demo
}

message SubmitPromptAck {
  bool ok = 1;
  string msg = 2;
}

message RunInferenceReq {
  string session_id = 1;
  int32 max_new_tokens = 2;
  float temperature = 3;
  int32 top_k = 4;
}

message Token {
  int64 token_id = 1;
  string text = 2;
}

message StepTiming {
  int64 step_index = 1;
  double ms = 2;
}

message Log {
  string level = 1;
  string message = 2;
}

message InferenceEvent {
  oneof evt {
    Token token = 1;
    StepTiming timing = 2;
    Log msg = 3;
  }
}

message Metrics {
  double tokens_per_second = 1;
  uint64 triple_pool_remaining = 2;
}
